src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,148 @@
+
+        public Node(Node<T> prev, T value, Node<T> next) {
+            this.prev = prev;
+            this.value = value;
+            this.next = next;
+        }
+    }
+
     @Override
     public void add(T value) {
+        if (first == null) {
+            first = new Node<>(null, value, null);
+        } else if (last == null) {
+            last = new Node<>(first, value, null);
+            first.next = last;
+        } else {
+            last.next = new Node<>(last, value, null);
+            last = last.next;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndexToBounds(index, size);
+        Node<T> node = getNodeByIndex(index);
+        if (node == null) {
+            add(value);
+            return;
+        }
+        node.prev = new Node<>(node.prev, value, node);
+        if (node.prev.prev == null) {
+            first = node.prev;
+        } else {
+            node.prev.prev.next = node.prev;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T listElement : list) {
+            add(listElement);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndexToBounds(index, size - 1);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        T oldValue = get(index);
+        getNodeByIndex(index).value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndexToBounds(index, size - 1);
+        Node<T> node = getNodeByIndex(index);
+        if (node.prev == null && node.next == null) {
+            first = null;
+            size--;
+            return node.value;
+        }
+        if (node.prev == null) {
+            node.next.prev = null;
+            first = node.next;
+            size--;
+            return node.value;
+        }
+        if (node.next == null) {
+            node.prev.next = null;
+            last = node.prev;
+            size--;
+            return node.value;
+        }
+        node.prev.next = node.next;
+        node.next.prev = node.prev;
+        size--;
+        return node.value;
     }
 
     @Override
     public boolean remove(T object) {
+        int index = getIndexByValue(object);
+        if (index >= 0) {
+            remove(index);
+            return true;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndexToBounds(int index, int size) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index is out of bounds!");
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
```

Comment:
[checklist](https://mate-academy.github.io/jv-program-common-mistakes/java-core/collections/linked-list.html): Don’t iterate your nodes from the head to tail or vice versa.